syntax = "proto2";

import "src/proto/vec3.proto";
package battle;

// A list of possible actions an agent can take. Some of these haven't
// been implemented (or flushed out yet)
message Actions {
  message Shoot {
    optional bool shoot = 1; 
  }
  message Rotate {
    optional state.Vec3 v = 1;
    optional double a = 2;
  }
  message Move {
    optional state.Vec3 target = 1;
  }
  message Crouch {
  }
  message Jump {
  }
  message ChangeWeapon {
  }
  optional Move move = 1;
  optional Rotate rotate = 2;
  optional Crouch crouch = 3;
  optional Jump jump = 4;
  optional ChangeWeapon change_weapon = 5;
  optional Shoot shoot = 6;
}

enum StatusCode {
  STATUS_OK = 1;
  STATUS_INVALID_ACTION = 2;
  STATUS_NO_OP = 3;
}

// The response of the game server to a desired action containing
// the result of the intended action.
message ActionResponse {
  message Result {
    optional StatusCode code = 1;
    optional string message = 2;

    message Shoot {
    }
    message Rotate {
      optional state.Vec3 v = 1;
      optional double a = 2;
    }
    message Move {
      optional state.Vec3 pos = 1;
    }
    message Crouch {
    }
    message Jump {
    }
    message ChangeWeapon {
    }
    oneof details {
      Move move = 3;
      Rotate rotate = 4;
      Crouch crouch = 5;
      Jump jump = 6;
      ChangeWeapon change_weapon = 7;
      Shoot shoot = 8;
    }
  }

  repeated Result results = 1;
}
